### ISSUE TITLE
Suggestion: Compatibility of st_searchbox with Streamlit 1.37

### ISSUE THREAD

**Issue by salmanrazzaq-94** [m-wrzr/streamlit-searchbox#54]
Hi, thank you for the great work. I am encountering compatibility issues with the st_searchbox component in Streamlit version 1.37. The current implementation causes the entire page to rerun, resulting in iframe distortion. To address this, I propose using the new "fragment" scope feature introduced in Streamlit 1.37 (https://docs.streamlit.io/develop/api-reference/execution-flow/st.fragment), which allows partial updates to the page, preventing unnecessary re-renders.

Proposed Solution:

Parameter Addition: 
- Introduce a rerun_scope parameter to the st_searchbox function and pass it to _process_search. The scope can be set to "app" (default) for full page reruns or "fragment" for partial updates. (https://docs.streamlit.io/develop/api-reference/execution-flow/st.rerun)

Implementation: 
- Modify st_searchbox and _process_search to handle the scope parameter and adjust rerun behavior accordingly.

Additional Consideration:
- Check the Streamlit version and rerun appropriately to ensure backward compatibility.

---
Solution #55

**Comment by salmanrazzaq-94** [comment]
@m-wrzr  Any update on this? This is critical for my app. Would really appreciate if you look into it.

**Comment by m-wrzr** [comment]
Hey, thanks for the PR! 🙂 I won't have time to look at it properly before Sunday, so can you just install it from directly from your fork in the meantime? Since fragments are quite new, I want to make sure that every that everything works as expected, I haven't used them before..

**Comment by salmanrazzaq-94** [comment]
Yes, definitely. Please take your time. The performance has improved significantly for me using fragments.

**Comment by m-wrzr** [comment]
Can you try to install it from here? `pip install -i https://test.pypi.org/simple/ streamlit-searchbox==0.1.15a1`

I just built & published the fork from @salmanrazzaq-94 as-is on test.pypi

Still wanted to wait for updates in https://github.com/streamlit/streamlit/issues/9002 and streamlit version `1.38` that would address this issue... but I guess if nothing happens soon I'll replace the strict version checks with warnings

**Comment by msr-agentsmyth** [comment]
@m-wrzr  Thank you for your valuable contributions to the community. I appreciate the great work you're doing and would love to implement any suggestions you have for improving the code.

I have another suggestion: could we incorporate a debouncing mechanism? Currently, the React component sends a function call with every keystroke, which can lead to delays and distortions, especially since API responses can sometimes be slow. By implementing debouncing, we could ensure that the API call is only made after the user has paused typing for a set period, such as 500ms. This would help avoid unnecessary API calls during fast typing, optimizing overall performance.

**Comment by salmanrazzaq-94** [comment]
@m-wrzr Thank you for your valuable contributions to the community. I appreciate the great work you're doing and would love to implement any suggestions you have for improving the code.

I have another suggestion: could we incorporate a debouncing mechanism? Currently, the React component sends a function call with every keystroke, which can lead to delays and distortions, especially since API responses can sometimes be slow. By implementing debouncing, we could ensure that the API call is only made after the user has paused typing for a set period, such as 500ms. This would help avoid unnecessary API calls during fast typing, optimizing overall performance.

**Comment by sw8989** [comment]
Thanks a lot @m-wrzr, that did the trick!

**Comment by m-wrzr** [comment]
@salmanrazzaq-94 great suggestion, thanks. I've added a debounce option to the next version. 👍

This doesn't help with the rerun/reset issue since it's mostly dependent on the time between script start & rerun (e.g. longer search functions don't have this issue). Added an optional extra delay option there, since I don't think there are any fixes coming soon... Not sure about any default values for this and the debounce in streamlit `1.37` though...

The fragment support is also merged into https://github.com/m-wrzr/streamlit-searchbox/pull/56 - I'll only get around to properly test everything in a few days, but if you want you wan try it out here https://test.pypi.org/project/streamlit-searchbox/0.1.15rc0/

Thanks again!
